### PR TITLE
Send globals along with `run_function` step definition

### PR DESCRIPTION
**Pull request by aksh-at**
Allows us to fix the issue where changing a global variable referenced in a function doesn't rebuild the function:

```python
BASE_MODEL = "foo"

def download_models():
    from transformers import LlamaForCausalLM, LlamaTokenizer

    LlamaForCausalLM.from_pretrained(BASE_MODEL)
    LlamaTokenizer.from_pretrained(BASE_MODEL)
```

We're filtering out any functions referenced here because they can't be serialized in a "stable" way, but there might be other types we should filter out. 

An alternative would be letting users specify parameters explicitly:

```python
def download_models(model):
    from transformers import LlamaForCausalLM, LlamaTokenizer

    LlamaForCausalLM.from_pretrained(model)
    LlamaTokenizer.from_pretrained(model)

image = modal.Image.debian_slim().run_function(download_models, "foo")
```

This isn't great because it's easy for users to miss, and then leads to the same confusion as now when changing a variable doesn't change the image.

### CODE REVIEW

**Comment by aksh-at** [comment]
@prbot approve

**Comment by ekzhang** [comment]
Can look on Monday?

**Comment by ekzhang** [comment]
Not sure if you want another pair of eyes or just want to merge immediately

**Comment by aksh-at** [comment]
No rush. I decided to hold off to think about this a bit more.

File: modal/functions.py
Comment (erikbern) [inline modal/functions.py:967]: Is this basically the same as inspect.getclosurevars?

Minor preference for putting this in function_utils.py maybe?

File: modal_proto/api.proto
Comment (erikbern) [inline modal_proto/api.proto:863]: Can we just serialize this into the previous field? Since I think we already do that client side with other arguments

**Review by erikbern — APPROVED** [review]

File: modal/functions.py
Comment (aksh-at) [reply modal/functions.py]: `inspect.getclosurevars` has some weird behavior, e.g. see this issue https://github.com/python/cpython/issues/80878:

> We are taking names from func.__code__.co_names, which also contains names of object's attributes we visited. As a result, we may get more names unexpectedly in 'globals', 'builtins', and even 'unbound'.

Makes sense to move it to function_utils.py, moved

File: modal_proto/api.proto
Comment (aksh-at) [reply modal_proto/api.proto:863]: Was going to do that earlier but I thought it'd be nice to be able to "turn off" using `build_function_globals` in the image hash if it turned out to have issues. Also think we should separate out the other arguments at some point into a separate PB as we discussed

File: modal/image.py
Comment (erikbern) [inline modal/image.py:200]: Minor thing (because this is some cruft I introduced) but we should probably replace `_get_info` with a property instead. In general let's avoid accessing internals of other objects.

But I can fix separately since it's mostly my fault that the `Function` interface is a bit janky

**Pull request merged**
